File: T1.SqlSharp/T1.SqlSharp/Expressions/SqlTopClause.cs
using System.Text;

namespace T1.SqlSharp.Expressions;

public class SqlTopClause : ISqlExpression
{
    public SqlType SqlType => SqlType.TopClause;
    public TextSpan Span { get; set; } = new();
    public required ISqlExpression Expression { get; set; }
    public bool IsPercent { get; set; }
    public bool IsWithTies { get; set; }
    public string ToSql()
    {
        var sql = new StringBuilder();
        sql.Append($"TOP {Expression.ToSql()}");
        if (IsPercent)
        {
            sql.Append(" PERCENT");
        }
        if(IsWithTies)
        {
            sql.Append(" WITH TIES");
        }
        return sql.ToString();
    }
}

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlTopClause.ts
import { SqlType } from './SqlType';
import { TextSpan } from '../StringParser';
import { ISqlExpression, SqlVisitor } from './ISqlExpression';

export class SqlTopClause implements ISqlExpression {
    SqlType: SqlType = SqlType.TopClause;
    Span: TextSpan = new TextSpan();
    Expression!: ISqlExpression;
    IsPercent: boolean = false;
    IsWithTies: boolean = false;

    Accept(visitor: SqlVisitor): void {
        visitor.Visit_TopClause(this);
    }

    ToSql(): string {
        let sql = `TOP ${this.Expression.ToSql()}`;
        if (this.IsPercent) {
            sql += ' PERCENT';
        }
        if (this.IsWithTies) {
            sql += ' WITH TIES';
        }
        return sql;
    }
} 